wireapp/kalium
# fix(dev): logout properly after expired session [AR-3307]

----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

After the session is expired, when combined with removal of the client, we get failures when trying to decode messages.

### Causes (Optional)

When the session expires, the account is only marked as logged out, but nothing else related to logging out is being done, client's data is not being cleared and the current client id is still kept which causes troubles.
After error saying that the client used is invalid or wrong event was used to fetch notifications (404), the slow sync is reset and restarted but incremental sync doesn't start which makes the app being stuck.

### Solutions

Refactor `UserSessionScopeProvider` and move `SessionManager` creation and all other code that's common for every platform directly to the `UserSessionScope`, to be also able to use `LogoutUseCase` in the `SessionManager` and logout the user properly when the `SESSION_EXPIRED` occurs.
Retry the incremental sync callback (listen for the completion of slow sync and retry incremental sync after) when the slow sync is restarted for recovery after the app gets a response with 404 meaning that wrong event or client id was used.
Clear the outdated client data when the app discovers that the retained client is no longer registered so that we don't keep the invalid "current client id" and "last processed event id" for that client.

### Testing

#### How to Test

Expire the session by for instance changing tokens in shared prefs, remove the current client from another device and open the app to see how it behaves.

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
